### PR TITLE
core: add ability to create hard-forks

### LINKED ISSUE #2464: Add hardforks and raise the fee of CreateMultisigAccount call
neo-project/neo#2712

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Close #2464.

### CODE REVIEW

File: pkg/core/interop/context.go [hunk 1]
```diff
@@ -368,3 +368,15 @@ func (ic *Context) GetBlock(hash util.Uint256) (*block.Block, error) {
 	}
 	return block, nil
 }
+
+// IsHardforkEnabled tells whether specified hardfork enabled at the current context height.
+func (ic *Context) IsHardforkEnabled(hf config.Hardfork) bool {
+	cfg := ic.Chain.GetConfig()
+	hfStr := hf.String()
+	for hf, height := range cfg.Hardforks {
```
Comment (roman-khimov) [inline pkg/core/interop/context.go:376]: OK until we have more than 10 of them.

File: pkg/core/blockchain.go [hunk 2]
```diff
@@ -259,6 +259,10 @@ func NewBlockchain(s storage.Store, cfg config.ProtocolConfiguration, log *zap.L
 		cfg.NativeUpdateHistories = map[string][]uint32{}
 		log.Info("NativeActivations are not set, using default values")
 	}
+	if cfg.Hardforks == nil {
+		cfg.Hardforks = map[string]uint32{}
+		log.Info("Hardforks are not set, using default value")
```
Comment (roman-khimov) [inline pkg/core/blockchain.go:264]: The default is to enable all of them, that's critical.

Comment (AnnaShaleva) [reply pkg/core/blockchain.go:264]: Sure, see the implementation of `IsHardforkEnabled`. If the hardfork is not in the dictionary, then it's enabled. The behaviour matches the C#'s one.

File: pkg/config/hardfork.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+//go:generate stringer -type=Hardfork
+
+// Hardfork represents the application hard-fork identifier.
+type Hardfork byte
+
+const (
+	// HF_2469_FixSyscallFees represents hard-fork introduced in #2469 changing the
+	// prices of System.Contract.CreateStandardAccount and
+	// System.Contract.CreateMultisigAccount interops.
+	HF_2469_FixSyscallFees Hardfork = 1 << iota // TODO: in C# the name is HF_2712_FixSyscallFees
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/config/hardfork.go]: Fixed.

File: pkg/core/interop/context.go [hunk 4]
```diff
@@ -368,3 +368,15 @@ func (ic *Context) GetBlock(hash util.Uint256) (*block.Block, error) {
 	}
 	return block, nil
 }
+
+// IsHardforkEnabled tells whether specified hardfork enabled at the current context height.
+func (ic *Context) IsHardforkEnabled(hf config.Hardfork) bool {
+	cfg := ic.Chain.GetConfig()
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/core/interop/context.go]: Fixed.

File: pkg/core/interop/context.go (hunk 1 above)
Comment (AnnaShaleva) [reply pkg/core/interop/context.go]: Don't know why I wrote this code... We can easily take element from map without this `range`, fixed.

**Comment by AnnaShaleva** [comment]
Tests are fixed.

File: pkg/core/interop/context.go [hunk 5]
```diff
@@ -368,3 +371,12 @@ func (ic *Context) GetBlock(hash util.Uint256) (*block.Block, error) {
...
 	return block, nil
 }
+
+// IsHardforkEnabled tells whether specified hard-fork enabled at the current context height.
+func (ic *Context) IsHardforkEnabled(hf config.Hardfork) bool {
+	height, ok := ic.Hardforks[hf.String()]
+	if ok {
+		return ic.BlockHeight() >= height
+	}
+	return true // Enable each hard-fork by default.
```
Comment (roman-khimov) [inline pkg/core/interop/context.go:381]: `return len(ic.Hardforks) == 0`?

File: config/protocol.unit_testnet.single.yml [hunk 6]
```diff
@@ -20,6 +20,8 @@ ProtocolConfiguration:
     RoleManagement: [0]
     OracleContract: [0]
     Notary: [0]
+  Hardforks:
```
Comment (roman-khimov) [inline config/protocol.unit_testnet.single.yml:23]: Hmm, but we have not documented it! And it's not an easy one.

Comment (AnnaShaleva) [reply config/protocol.unit_testnet.single.yml:23]: Not sure, should we explain why hardforks have been added?

Comment (roman-khimov) [reply config/protocol.unit_testnet.single.yml:23]: The concept itself is pretty well-known, but I'll take at look at the text.

File: docs/node-configuration.md [hunk 7]
```diff
@@ -200,31 +200,31 @@ where:
...
+| CommitteeHistory                | map[uint32]int        | none                                                                                                                                                                                   | Number of committee members after the given height, for example `{0: 1, 20: 4}` sets up a chain with one committee member since the genesis and then changes the setting to 4 committee members at the height of 20. `StandbyCommittee` committee setting must have the number of keys equal or exceeding the highest value in this option. Blocks numbers where the change happens must be divisible by the old and by the new values simultaneously. If not set, committee size is derived from the `StandbyCommittee` setting and never changes.                                                                                                                                                  |
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply docs/node-configuration.md]: Fixed.

**Pull request merged**